service/history/ndc_task_util.go

```diff
@@ -146,11 +149,23 @@ func LoadMutableStateForTask(
 	logger log.Logger,
 ) (workflow.MutableState, error) {
 
+	if err := validateTaskByClock(shardContext, task); err != nil {
+		return nil, err
+	}
+
 	mutableState, err := wfContext.LoadMutableState(ctx, shardContext)
 	if err != nil {
 		return nil, err
 	}
 
+	// TODO: With validateTaskByClock check above, we should never run into the situation where
+	// mutable state cache is stale. This is based on the assumption that shard context
+	// will never re-acquire the shard after it has been stolen.
+	// We should monitor the StaleMutableStateCounter metric and remove the logic below once we are confident.
+
+	// Validation based on eventID is not good enough as certain operation does not generate events.
+	// For example, scheduling transient workflow task, or starting activities that have retry policy.
+
```

Comment:
Agree this is redundant but we still need a check that the task is not stale and a namespace failover version check.
The namespace failover check may either invalidate mutable state or the task:
https://github.com/temporalio/temporal/pull/5356/files#diff-4dd87cd5210fc76333fb926a6d3b26ebc1aba6069fd1fe173953df39f17bf0b9R145